walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,70 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Set;
+import java.util.List;
+import java.util.ArrayList;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> times = new ArrayList<>();
+    if (request.getDuration() > 1440) {
+      return times;
+    }
+    //asume all time is valid
+    times.add(TimeRange.WHOLE_DAY);
+    for (Event event : events) {
+      if (hasOverlappingAttendendees(event.getAttendees(), request.getAttendees())) {
+        TimeRange eventTime = event.getWhen();
+        for (int i = 0; i < times.size(); i++) {
+          TimeRange openTime = times.get(i);
+          //found an overlapping time in free time
+          if (eventTime.overlaps(openTime)) {
+            //four cases of overlapping
+            if (eventTime.contains(openTime.start()) && eventTime.contains(openTime.end())) {
+              //eventTime:
+              //openTime is completely inside evenTime -> remove it from the free times
+              times.remove(i--);
+            } else if (openTime.contains(eventTime.start()) && openTime.contains(eventTime.end())) {
+              //eventTime is completely inside openTime -> split openTime into two
+              //evenTime:        |-|
+              //openTime:     |---------|
+              times.remove(i--);
+              times.add(TimeRange.fromStartEnd(openTime.start(), eventTime.start(), false));
+              times.add(TimeRange.fromStartEnd(eventTime.end(), openTime.end(), false));
+            } else if (eventTime.contains(openTime.start())) {
+              //eventTime:     |-----|
+              //openTime:           |-----|
+              //-> remove overlapping beginning of openTime
+              times.set(i, TimeRange.fromStartEnd(eventTime.end(), openTime.end(), false));
+            } else {
+              //eventTime:       |-----|
+              //openTime:     |-----|
+              //-> remove overlapping end of openTime
+              times.set(i, TimeRange.fromStartEnd(openTime.start(), eventTime.start(), false));
+            }
+          }
+        }
+      }
+    }
+    //remove time slots that are too short
+    for (int i = 0; i < times.size(); i++) {
```

Comment:
same here where a for loop that involves decrementing is probably better off with a queue.